pkg/pillar/cmd/zedagent/parseconfig.go

```diff
@@ -1383,6 +1383,77 @@ func parseIpspec(ipspec *zconfig.Ipspec,
 		config.DhcpRange.Start = start
 		config.DhcpRange.End = end
 	}
+
+	// take some default
+	if config.DhcpRange.Start == nil || config.DhcpRange.End == nil ||
+		config.Gateway == nil {
+		addrCount := types.GetIPAddrCountOnSubnet(config.Subnet)
+		if addrCount == 0 {
+			return nil
+		}
+		base := types.GetIPSubnetBase(config.Subnet.IP, config.Gateway)
+		ipStart := 1
+		if base != 0 {
+			ipStart = base
+		}
+		dhcpRangeStart := ipStart + 1
+		if addrCount >= types.LargeAppCount {
+			// when gateway is less than the half point,
+			// the dhcpRange starts at the half point,
+			// provided the DhcpRange.End is not set
+			if base < addrCount/2 &&
+				config.DhcpRange.End == nil {
+				dhcpRangeStart = addrCount / 2
+			}
+		}
+		// last addressable endpoint, with 0 base, and subnet.IP as start,
+		// it accounts for (2^(iplen - subnetMask) - 2) addresses
+		dhcpRangeEnd := addrCount - 2
```

Comment:
dhcpRangeEnd can be negative for tiny subnets.
Must better to have a check up front on line 1391 to check that there is some minimal size of the subnet (16 addresses sounds reasonable to me) for and if not generate an error indicating that the dhcp range is not set and it can not be defaulted since the subnet is so small.

Reply:
With that check and avoid special handling of the gateway in this default code, then all you need is a check that if the dhcp range is not set, it defaults to start = types.AddToIP(config.Subnet.IP, addrCount / 2) and end = types.AddToIP(config.Subnet.IP, addrCount)

Your code below will adjust the end if it includes the broadcast address.